### ISSUE TITLE
New axis: piecewise_regular

### ISSUE THREAD

**Issue by HDembinski** [HDembinski/histogram#47]
A composed axis of several connected regular axis without gaps. Should use regular_axis underneath.

**Comment by swornim1** [comment]
I would like to work on this project. Please explain the further details required to accomplish the task. Thank you.

**Comment by HDembinski** [comment]
Hi,

thank you for your interest.

you should look into the implementation of boost::histogram::axis::regular to get an idea how this axis would work, also have a look at the Axis concept in the documentation.

I gave this axis more thought in the meantime and it should be even more generalized: piecewise is a new axis which pieces several other axis types together.

```
template <class... Ts>
class piecewise { ... };
```

The `Ts...` are other axis types.

It is created by  a `make_piecewise` factory function.

```
template <class... Ts>
piecewise<Ts...> make_piecewise(Ts... ts) {
    ...
}
```

Upon construction, `piecewise` checks that the boundaries of its axis arguments are connected, meaning that the uppermost bin edge of the first axis is equal to the lowermost bin edge of the second axis, etc. It keeps a std::tuple of the axis types. When a value is passed to the index method, it looks the sub-axis up whose range covers the argument value and passes the value to index method of that sub-axis. The index returned is shifted by the number of bins left of that sub-axis.

**Comment by ryanelandt** [comment]
I can see how piecing together floating axis types might give spacing that is difficult to achieve with a simple transformation (e.g., sqrt, log, etc). But, I'm not seeing why one would want to stitch together integer axes. If I understand it correctly, the width of an integer axis is always one. Is there a potential advantage?

**Comment by HDembinski** [comment]
> But, I'm not seeing why one would want to stitch together integer axes.

True, it makes no sense if we do not allow gaps between consecutive axes. I had not read the whole issue again to refresh my memory, so I was considering to allow gaps between axes. In that case it might be useful to stitch several integer axes together with gaps between them.

You could use a category<int> axis to have gaps in an integer range, but bin lookup for a category<int> axis is more expensive and you need to allocate memory from the heap for a category axis, as opposed to an integer axis.
